### PR TITLE
V4 utils

### PR DESCRIPTION
**Pull request by jshrt**
* refactored normalizeSetArguments to accept a data argument to be all types except function or undefined 
* added normalizeSetArguments test

### CODE REVIEW

File: src/util/utils.ts [hunk 1]
```diff
@@ -181,65 +181,87 @@ export const getUid = (): string => {
...
-      }
-      if (result.callback !== undefined && typeof result.callback !== 'function') {
-        throw Error ('Invalid set callback argument')
-      }
-      if (result.data === undefined) {
-        throw Error ('Invalid set data argument')
-      }
-      return result
+    if (!isPath(args[startIndex])) {
+      result.callback = isCallback(args[startIndex + 1]) ? args[startIndex + 1] : false
```
Comment (yasserf) [inline src/util/utils.ts]: Why is callback a function or false?

File: src/util/utils.ts [hunk 2]
```diff
@@ -181,65 +181,87 @@ export const getUid = (): string => {
...
     }
+  }
 
-    throw Error ('Invalid set arguments')
+  if (args.length === startIndex + 3) {
+    result = {
+      path: isPath(args[startIndex]) ? args[startIndex] : false,
```
Comment (yasserf) [inline src/util/utils.ts]: How is path false?

Comment (jshrt) [reply src/util/utils.ts]: those `false` values in this case are a way to trigger precise error messages, as `undefined` can be an acceptable case for args.length cases

**Comment by jshrt** [comment]
The refactor takes into consideration two use cases:
1. When data is supposed to be at root level of the record, which then only allows it to be of type `object`
2. When data is at a certain path in the record, which then allows it to be of type `object` and also primitives. (_as discussed last time_)

nested method calls feels like too much, but seemed more readable than redundant type checking.

Let me know what to keep and to discard

File: test/util/utilsSpec.ts [hunk 3]
```diff
@@ -576,4 +622,9 @@ describe('parseUrl adds all missing parts of the url', () => {
     expect(utils.parseUrl('localhost?query=value#login', '/deepstream'))
       .to.equal('ws://localhost/deepstream?query=value#login')
   })
+
+  it('rejects urls with @no-host', () => {
```
Comment (yasserf) [inline test/util/utilsSpec.ts]: I would stop using @ notations in tests, never seen that standard anwhere before.

File: src/util/utils.ts (hunk 2 above)
Comment (yasserf) [reply src/util/utils.ts]: Cool, makes sense now

File: test/util/utilsSpec.ts (hunk 3 above)
Comment (jshrt) [reply test/util/utilsSpec.ts]: I agree.

Those are basically meaningless leftovers that I forgot to remove, I was using them with `--grep` when I was focusing on these cases..

**Pull request merged**